### PR TITLE
Added categorical encoding function

### PR DESCRIPTION
**Pull request by Nandha951**
## **Description**
This pull request introduces a new **categorical encoding** feature to the project. The function supports **One-Hot Encoding** and **Ordinal Encoding** for categorical variables, allowing users to efficiently transform categorical data into numerical formats. This addition is designed to enhance data preprocessing capabilities within the framework.

The changes include:
1. A new Python file `categorical_encoding.py` under the `src/koheesio/` directory.
2. Implementation of a flexible `categorical_encoding` function.
3. Addition of a `EncodingConfig` class for user-configurable options (e.g., encoding type).
4. Comprehensive unit tests under `tests/test_categorical_encoding.py`.

---

## **Related Issue**
This pull request addresses the need for robust categorical data transformation functionality as identified in the enhancement discussions. (Link to any related issue or discussion if applicable, or mention "N/A" if there isn't one.)

---

## **Motivation and Context**
This change is required to handle categorical data during data preprocessing for machine learning or analytics workflows. The new feature provides:
- A **One-Hot Encoding** option for creating binary columns for each category while avoiding the dummy variable trap.
- An **Ordinal Encoding** option to assign numerical values to categories, useful for models that can process ordinal relationships.

This functionality improves the versatility and usability of the Koheesio framework in real-world scenarios.

---

## **How Has This Been Tested?**
The implementation was tested using unit tests created in `tests/test_categorical_encoding.py`. The tests include:
1. **One-Hot Encoding**:
   - Validated the creation of binary columns for multiple categorical variables.
   - Verified the exclusion of original categorical columns.
2. **Ordinal Encoding**:
   - Ensured that the correct numerical mappings were assigned to categories.
   - Checked for consistency when applied to multiple variables.
   
All tests were executed in the local environment using Python's unittest framework, and they passed successfully without affecting other parts of the codebase.

---

## **Screenshots (if appropriate):**
N/A

---

## **Types of Changes**
- [x] New feature (non-breaking change which adds functionality)

---

## **Checklist**
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: src/koheesio/categorical_encoding.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
+# Import necessary libraries
+from sklearn.preprocessing import OneHotEncoder
```
Comment (dannymeijer) [inline src/koheesio/categorical_encoding.py:2]: You're introducing a new dependency here, this should be included in the pyproject.toml.

File: src/koheesio/categorical_encoding.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+import pandas as pd
+from pydantic import BaseModel
+
+# Define the EncodingConfig class
+class EncodingConfig(BaseModel):
+    drop_first: bool = True  # Whether to drop the first dummy column
+
+# Define the categorical_encoding function
+def categorical_encoding(data, columns, config: EncodingConfig):
```
Comment (dannymeijer) [inline src/koheesio/categorical_encoding.py:11]: Please adjust this to become a `Step` class to take full advantage of Koheesio's built-in type checking.

Arguably, this should become a Pandas version of the Transformation class we have in the Spark module.

File: src/koheesio/categorical_encoding.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    Encodes categorical columns using one-hot encoding.
+
+    Args:
+        data (pd.DataFrame): Input dataset.
+        columns (list): List of categorical columns to encode.
+        config (EncodingConfig): Configuration for encoding.
+
+    Returns:
+        pd.DataFrame: Dataset with one-hot encoded columns.
+    """
```
Comment (dannymeijer) [inline src/koheesio/categorical_encoding.py:22]: Docstrings are supposed to be in Numpy notation

File: src/koheesio/categorical_encoding.py
Comment (dannymeijer) [inline src/koheesio/categorical_encoding.py:1]: This file should be placed in the proper module. If you intent this to be a pandas based Step, please put it in the pandas module.

**Review by dannymeijer — CHANGES_REQUESTED** [review]
First of all, thank you for your contribution, it's very much appreciated.

Please see the comments I gave above.

A few things:
- Your code should be adjusted to use Koheesio Step class, preferably a pandas adjusted version of the Transformation class we have in the spark module
- Your code should be moved to an appropriate module
- The extra dependency you're introducing is not part of the pyproject atm
- Also, I would like you to add extra documentation (module docstring) to explain your usecase; add some examples as well once you've adjusted your code

I would love to discuss with you the intend of what you are trying to achieve. Feel free to reach out in a DM/email - my contact information is in my profile (LinkedIn, email).

File: src/koheesio/categorical_encoding.py (hunk 1 above)
Comment (Nandha951) [reply src/koheesio/categorical_encoding.py]: I have added dependency in the pyproject.toml

File: src/koheesio/categorical_encoding.py (hunk 2 above)
Comment (Nandha951) [reply src/koheesio/categorical_encoding.py]: Adjusted to using the Step class

File: src/koheesio/categorical_encoding.py (hunk 3 above)
Comment (Nandha951) [reply src/koheesio/categorical_encoding.py]: Modified Docstrings to be in Numpy notation

File: src/koheesio/categorical_encoding.py
Comment (Nandha951) [reply src/koheesio/categorical_encoding.py:1]: Placed the file in the pandas module.

File: pyproject.toml [hunk 4]
```diff
@@ -41,6 +41,8 @@ dependencies = [
   "pytz>=2023.3",
   "pyyaml>=6.0",
   "tomli>=2.0.1",
+  "pandas>=1.5.0",
+  "scikit-learn>=1.2.0"
```
Comment (dannymeijer) [inline pyproject.toml:45]: I don't want to make these to level dependencies + we already have pandas as an extra dependency.
Let's make an extra called "ml" and put scikit-learn in there. That way you can install the extra dependency as `koheesio[pandas,ml]`

File: src/koheesio/pandas/categorical_encoding.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+import pandas as pd
+from pydantic import BaseModel
+from typing import List, Optional
+from koheesio.steps import Step
+
+from typing import List, Dict
+import pandas as pd
+from pydantic import BaseModel
+
+class PandasCategoricalEncoding(BaseModel):
```
Comment (dannymeijer) [inline src/koheesio/pandas/categorical_encoding.py:11]: make this class a `PandasStep` : `from koheesio.pandas import PandasStep`

File: src/koheesio/pandas/categorical_encoding.py [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+    encoding_type : str, optional
+        Type of encoding, either "one-hot" or "ordinal" (default is "one-hot").
+    drop_first : bool, optional
+        Whether to drop the first dummy column for one-hot encoding (default is True).
+    ordinal_mapping : dict, optional
+        A dictionary mapping categorical values to integers for ordinal encoding.
+    """
+
+    columns: List[str]
+    encoding_type: str = "one-hot"
```
Comment (dannymeijer) [inline src/koheesio/pandas/categorical_encoding.py:28]: change the type to `Literal["one-hot", "ordinal"]`, that way you don't need the extra check you put in the `__init__` method

File: src/koheesio/pandas/categorical_encoding.py [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+    drop_first: bool = True
+    ordinal_mapping: Dict[str, Dict] = None
+
+    def __init__(self, **kwargs):
```
Comment (dannymeijer) [inline src/koheesio/pandas/categorical_encoding.py:32]: This becomes obsolete if you make the type a Literal as stated above

File: src/koheesio/pandas/categorical_encoding.py [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+    drop_first : bool, optional
+        Whether to drop the first dummy column for one-hot encoding (default is True).
+    ordinal_mapping : dict, optional
+        A dictionary mapping categorical values to integers for ordinal encoding.
+    """
+
+    columns: List[str]
+    encoding_type: str = "one-hot"
+    drop_first: bool = True
+    ordinal_mapping: Dict[str, Dict] = None
```
Comment (dannymeijer) [inline src/koheesio/pandas/categorical_encoding.py:30]: Make all these Fields, example:
```python
from koheesio.models import Field
...
class PandasCategoricalEncoding(PandasStep):

    columns: List[str] = Field(..., description="...")
    encoding_type: Literal["one-hot", "ordinal"] = Field(default="one-hot", description="...")
    ...
```

File: src/koheesio/pandas/categorical_encoding.py [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+        pd.DataFrame
+            The dataset with the specified categorical columns encoded.
+        """
+        if self.encoding_type == 'one-hot':
+            data = pd.get_dummies(data, columns=self.columns, drop_first=self.drop_first)
+        elif self.encoding_type == 'ordinal':
+            for column in self.columns:
+                if column in data.columns and self.ordinal_mapping and column in self.ordinal_mapping:
+                    data[column] = data[column].map(self.ordinal_mapping[column]).fillna(-1).astype(int)
+        return data
```
Comment (dannymeijer) [inline src/koheesio/pandas/categorical_encoding.py:57]: a few things about `execute` (for when you change to `Step`:
- execute takes no arguments, instead at the DataFrame as one of the input Fields; add something like this `df: Optional[pd.DataFrame] = Field(default=None, description="...")` - I will explain why you want this as Optional in a bit
- execute is expected to deal with input (your Fields) and generate Output
- this Output does not need to be returned explicitly, the Step parent-class takes care of this.
- instead, add a `.transform` method that can take a DataFrame as input

This means you can change your code like this:
1. add an `Output` class
2. add a `.transform` method
3. update your `execute` method accordingly

Should look something like this (of course add docstrings and things like that):
```python
class PandasCategoricalEncoding(PandasStep):
    ...
    class Output(PandasStep.Output):
        df: pd.DataFrame = Field(..., description="output pandas DataFrame"
    
    def transform(self, df: Optional[pd.DataFrame]):
        self.df = df or self.df
        if not self.df:
            raise RuntimeError("No valid Dataframe was passed")
        self.execute()
        return self.output.df

    def execute(self) -> Output:
        if self.encoding_type == 'one-hot':
            self.output.df = pd.get_dummies(self.df, columns=self.columns, drop_first=self.drop_first)
        elif self.encoding_type == 'ordinal':
            data = self.df
            for column in self.columns:
               for column in self.columns:
                if column in data.columns and self.ordinal_mapping and column in self.ordinal_mapping:
                    data[column] = data[column].map(self.ordinal_mapping[column]).fillna(-1).astype(int)
            self.output.df = data
```

This way you can interact with your class this way:
```python
encoding_step = PandasCategoricalEncoding(
    columns=["color"],
    encoding_type="one-hot",
    drop_first=False  # Adjusted to match expected columns
)
encoded_data = encoding_step.transform(self.data)
```

This way this interface matches what we do for Spark. 
Note: I will work on making Transformation base classes for Pandas in a separate PR.

For reference: https://github.com/Nike-Inc/koheesio/blob/9bd29ec553cc6306520926a89d5e6405641b5994/src/koheesio/spark/transformations/__init__.py#L35-L192

File: tests/test_categorical_encoding.py [hunk 10]
```diff
@@ -0,0 +1,81 @@
+import unittest
+import pandas as pd
+from src.koheesio.pandas.categorical_encoding  import PandasCategoricalEncoding
+import unittest
+import pandas as pd
+
+class TestPandasCategoricalEncoding(unittest.TestCase):
```
Comment (dannymeijer) [inline tests/test_categorical_encoding.py:7]: don't use unittest - use pytest.
1. get rid of the unittest.TestCase (just let it be a regular class)
2. change your `self.assert...` (from unittest) to regular python `assert` (this is how pytest works)
3. get rid of your setUp - just make the input dataframe a module level variable, OR make it a fixture (a bit overkill for your purpose here)
4. change your code to match the interface I proposed above

File: tests/test_categorical_encoding.py [hunk 11]
```diff
@@ -0,0 +1,81 @@
+import unittest
+import pandas as pd
+from src.koheesio.pandas.categorical_encoding  import PandasCategoricalEncoding
+import unittest
+import pandas as pd
```
Comment (dannymeijer) [inline tests/test_categorical_encoding.py:5]: you're importing pandas twice. Also, import pandas through the koheesio module (as stated above) to avoid conflict:
`from koheesio.pandas import pandas as pd`

File: src/koheesio/pandas/categorical_encoding.py [hunk 12]
```diff
@@ -0,0 +1,57 @@
+from sklearn.preprocessing import OneHotEncoder
+import pandas as pd
+from pydantic import BaseModel
+from typing import List, Optional
+from koheesio.steps import Step
+
+from typing import List, Dict
+import pandas as pd
```
Comment (dannymeijer) [inline src/koheesio/pandas/categorical_encoding.py:8]: import pandas through the koheesio module (as stated above) to avoid conflict:
`from koheesio.pandas import pandas as pd`

Please run isort (through `make fmt` or `hatch fmt,` or run `ruff`)

**Review by dannymeijer — CHANGES_REQUESTED** [review]
Starting to look really good! :) 
I left some detailed comments on what I'd like to see change

File: src/koheesio/pandas/categorical_encoding.py (hunk 8 above)
Comment (dannymeijer) [reply src/koheesio/pandas/categorical_encoding.py:30]: (and of course add appropriate description to each)

File: src/koheesio/pandas/categorical_encoding.py (hunk 9 above)
Comment (dannymeijer) [reply src/koheesio/pandas/categorical_encoding.py:57]: making `df` an `Optional` type allows us to either give the `df` as an argument when initializing the class, or pass it through `transform` - this is exactly how we do it inside the Spark module at the moment

**Comment by dannymeijer** [comment]
Since no response was provided since several weeks, I am closing this PR. Please re-open a new contribution request once you feel ready to do so and once the concerns have been addressed.
